### PR TITLE
Create an OPAL "core" library for internal usage

### PR DESCRIPTION
**Pull request by jjhursey**
* Fixes Issue https://github.com/open-mpi/ompi/issues/9869
 * Split `libopen-pal.so` into two libraries:
   - `libopen-pal_core.la` : Internal "core" portion of OPAL containing the essential source and MCA needed for mpicc/mpirun tools to link against. The "core" library is not installed.
   - `libopen-pal.la` : Includes "core" plus all of the other OPAL project sources. The `.so` version of this is installed.
 * The "core" library contains the following:
   - `opal/class`
   - `opal/mca/backtrace`
   - `opal/mca/dl`
   - `opal/mca/installdirs`
   - `opal/mca/threads`
   - `opal/mca/timer`
   - `opal/runtime/*_core.[c|h]`
   - `opal/runtime/opal_info_support.c`
   - `opal/util (most - see Makefile.am)`
   - `opal/util/keyval`
 * The "core" library is linked into the following tools instead of the full `libopen-pal.so`:
   - `ompi/tools/mpirun`
   - `ompi/tools/wrappers` (by extension of `opal/tools/wrappers`)
   - `opal/tools/wrappers`
 * The `opal/runtime` files were divided into a 'core' set vs 'main' set

### CODE REVIEW

**Comment by jjhursey** [comment]
This branch is functional and synced to `main`. I _think_ it is done, but I'm putting it as a WIP so I can bounce it off of CI to test configurations that I haven't tried.

The wrapper compilers and `mpirun` are all adjusted to use the `libopen-pal_core.la`. I did not work on `ompi_info` since it'll need to load the MCA components anyway to display their registered parameters.

**Comment by jjhursey** [comment]
Data point: Open MPI `main` compared to this PR. Linked in the UCX libraries (which carry dependencies).

Below are the results for `mpicc`, but `mpirun` and the other wrapper compilers are the same in their difference from `main`. No changes to the 3rd-party binaries (e.g., `hwloc-ps`, `pmixcc`, `prterun`). No change to `ompi_info`, as noted in the prior comment.

### Open MPI `main`

```
shell$ ldd ./bin/mpicc | sort
	/lib64/ld64.so.2 (0x00007fff98750000)
	libblkid.so.1 => /lib64/libblkid.so.1 (0x00007fff97330000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007fff974e0000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fff97860000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fff97a80000)
	libevent_core-2.1.so.6 => /lib64/libevent_core-2.1.so.6 (0x00007fff97dc0000)
	libevent_pthreads-2.1.so.6 => /lib64/libevent_pthreads-2.1.so.6 (0x00007fff97d90000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fff97400000)
	libhwloc.so.15 => /tmp/ompi-main-debug-orig/lib/libhwloc.so.15 (0x00007fff97ce0000)
	libmount.so.1 => /lib64/libmount.so.1 (0x00007fff97440000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fff97ba0000)
	libnuma.so.1 => /lib64/libnuma.so.1 (0x00007fff98300000)
	libopen-pal.so.0 => /tmp/ompi-main-debug-orig/lib/libopen-pal.so.0 (0x00007fff98570000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007fff971f0000)
	libpmix.so.0 => /tmp/ompi-main-debug-orig/lib/libpmix.so.0 (0x00007fff97e60000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fff98260000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fff98230000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fff972a0000)
	libucm.so.0 => /smpi_dev/jjhursey/local/ucx-1.11.2/lib/libucm.so.0 (0x00007fff982b0000)
	libucp.so.0 => /smpi_dev/jjhursey/local/ucx-1.11.2/lib/libucp.so.0 (0x00007fff98440000)
	libucs.so.0 => /smpi_dev/jjhursey/local/ucx-1.11.2/lib/libucs.so.0 (0x00007fff98340000)
	libuct.so.0 => /smpi_dev/jjhursey/local/ucx-1.11.2/lib/libuct.so.0 (0x00007fff983d0000)
	libudev.so.1 => /lib64/libudev.so.1 (0x00007fff97ab0000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007fff97e30000)
	libuuid.so.1 => /lib64/libuuid.so.1 (0x00007fff97300000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fff973c0000)
	linux-vdso64.so.1 (0x00007fff98730000)
```

### This PR

```
shell$ ldd ./bin/mpicc | sort
	/lib64/ld64.so.2 (0x00007fff9d8b0000)
	libblkid.so.1 => /lib64/libblkid.so.1 (0x00007fff9c900000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007fff9cab0000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fff9ce30000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fff9d050000)
	libevent_core-2.1.so.6 => /lib64/libevent_core-2.1.so.6 (0x00007fff9d390000)
	libevent_pthreads-2.1.so.6 => /lib64/libevent_pthreads-2.1.so.6 (0x00007fff9d360000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fff9c9d0000)
	libhwloc.so.15 => /tmp/ompi-main-debug/lib/libhwloc.so.15 (0x00007fff9d2b0000)
	libmount.so.1 => /lib64/libmount.so.1 (0x00007fff9ca10000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fff9d170000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007fff9c7c0000)
	libpmix.so.0 => /tmp/ompi-main-debug/lib/libpmix.so.0 (0x00007fff9d430000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fff9d830000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fff9d800000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fff9c870000)
	libudev.so.1 => /lib64/libudev.so.1 (0x00007fff9d080000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007fff9d400000)
	libuuid.so.1 => /lib64/libuuid.so.1 (0x00007fff9c8d0000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fff9c990000)
	linux-vdso64.so.1 (0x00007fff9d890000)
```

### Diff

A diff shows that the `mpicc` generated by this PR does not include the UCX libraries or `libopen-pal.so`
```
shell$ diff /tmp/mpicc-from-main.txt /tmp/mpicc-from-this-pr.txt 
12,13d11
< libnuma.so.1
< libopen-pal.so.0
19,22d16
< libucm.so.0
< libucp.so.0
< libucs.so.0
< libuct.so.0
```

**Comment by rhc54** [comment]
I honestly have forgotten - IIRC, your "mpirun" wrapper has a dependency on OPAL only for installdirs so it can properly deal with any prefix (envar or whatever) - is that correct? I'm not sure it all really matters, but you could probably strip things down even further for just the "mpirun" wrapper if you were so inclined. Installdirs shouldn't have much in the way of dependencies (certainly not hwloc and pmix, for example).

**Comment by bosilca** [comment]
#9893 ?

**Comment by bosilca** [comment]
I'm just trying to understand if there was a conceptual  issue with the approach proposed there.

**Comment by jjhursey** [comment]
Brian asked for help with his branch [here](https://github.com/open-mpi/ompi/issues/9869#issuecomment-1224225275), and I was asked to pick it up and drive it to completion for the v5 timeline. I saw some changes from George's branch were mixed in, so I assumed that Brian had already merged in the delta between the two approaches. I finished the work on Brian's branch, then squashed/rebased the branch to bring it up to the present. I added George and Brian as co-authors to the main commit due to their efforts in the branch up to the point where I picked it up.

The approach in this PR seems to fit nicely with the existing infrastructure. We could move some code around to make it a bit clearer which is "core" OPAL and which is "top/main" OPAL, but I'm hoping that the documentation (also a part of this PR) will cover that.

Note: This is still in a draft PR status because I wanted to bounce it off of CI. I'll work on bring it back for review later today.

File: opal/util/output.c [hunk 1]
```diff
@@ -63,6 +63,8 @@ static opal_output_stream_t verbose;
 static char *output_dir = NULL;
 static char *output_prefix = NULL;
 
+static opal_output_pmix_cleanup_fn_t output_pmix_cleanup_fn = NULL;
```
Comment (bosilca) [inline opal/util/output.c:66]: unlike `pmix`, `output` is now part of the core functionality. why are we using pmix in the name of this type in the `core` library ?

**Review by bosilca — APPROVED** [review]

File: opal/mca/if/configure.m4 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+dnl -*- autoconf -*-
```
[the remark this answers is not in the document]
Comment (bwbarrett) [reply opal/mca/if/configure.m4:1]: Laziness; I think we could hoist that out of core without breaking anything (I split the frameworks based on their current initialization in opal_init vs. opal_init_util, which was a good starting point but probably not right.

File: opal/util/output.c (hunk 1 above)
Comment (jjhursey) [reply opal/util/output.c:65]: That is correct. `output` is part of the OPAL core, but `pmix` is **not** part of the OPAL core. The `output.c` code had this code in it:
https://github.com/open-mpi/ompi/blob/b2c26c5993c895a8aab8b7d582aea9bcca9a89b8/opal/util/output.c#L786-L787

The call was inserted to tell the host RM not to clean up this file when cleaning up session directories and whatnot when the job completes. The `open_file` function may be called whenever a new framework/component opens a new output handle.

So I needed to find a way to get a pointer to the [`opal_pmix_register_cleanup()`](https://github.com/open-mpi/ompi/blob/main/opal/mca/pmix/base/pmix_base_fns.c#L460-L523) OPAL wrapper around the `PMIx_Job_control_nb ` PMIx function into the OPAL core without linking to it (so tools wouldn't complain about the missing symbol). 

To do this I added a function registration mechanism. I added the `pmix` name as a reminder of how it connects.

File: opal/util/output.c [hunk 3]
```diff
@@ -474,6 +478,22 @@ void opal_output_set_output_file_info(const char *dir, const char *prefix, char
     }
 }
 
+void opal_output_register_pmix_cleanup_fn(opal_output_pmix_cleanup_fn_t cleanup_fn)
```
[the remark this answers is not in the document]
Comment (jjhursey) [reply opal/util/output.c:480]: The `open_file` function might have been called in a few of the OPAL core frameworks/components before we can register the handler during the OPAL main initialization (which knows the PMIx symbol). This function first registers all of the previously opened files (if any) before setting the handle for any files to be opened later. It makes sure we cover them all.

File: opal/mca/if/configure.m4 (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjhursey) [reply opal/mca/if/configure.m4]: Good point. I just pushed on a new commit that moves the `if` framework from OPAL core to OPAL main/top. A quick build/run showed it worked out ok without too much trouble.

**Comment by jjhursey** [comment]
@artemry-nv  Mellanox CI failed on `mca_base_env_list` even after https://github.com/open-mpi/ompi/pull/10788#issuecomment-1245626168
```
+ mca='--mca pml ucx --mca btl ^vader,tcp,openib,uct'
+ '[' 2 -gt 0 ']'
+ echo 'mca_base_env_list=XXX_A=1;XXX_B=2;XXX_C;XXX_D;XXX_E'
++ /__w/1/ompi/ompi_install/bin/mpirun --mca pml ucx --mca btl '^vader,tcp,openib,uct' --np 2 --tune /__w/1/ompi/test_amca.conf /__w/1/jenkins_scripts/jenkins/ompi/env_mpi
++ grep --count '^XXX_'
--------------------------------------------------------------------------
WARNING: The format of "mca_base_env_list" parameter is a delimited list of VAR=VAL or
VAR instances. By default, the delimiter is a semicolon: VAR1=VAL1;VAR2;VAR3=VAL3;...
You can set other via "mca_base_env_list_delimiter" parameter. If the delimiter is a
semicolon, the value of "mca_base_env_list" variable should be quoted to not interfere
with SHELL command line parsing. In the case where a value is not assigned to variable
VAR, the value will be taken from the current environment.
The following environment variable was not found in the environment:
  Variable:             XXX_C
  MCA variable value:   XXX_A=1;XXX_B=2;XXX_C;XXX_D;XXX_E
--------------------------------------------------------------------------
--------------------------------------------------------------------------
No executable was specified on the prterun command line.

Aborting.
--------------------------------------------------------------------------
```

I think the next test down needs to comment out as well (for now, until it is fixed)
https://github.com/mellanox-hpc/jenkins_scripts/blob/1d66874bec127c22a2248b8d0da4cb5c7bdf5158/jenkins/ompi/ompi_test.sh#L236-L243

There also look to be some other cases that exercise `mca_base_env_list` that might need to be checked as well.

**Comment by jjhursey** [comment]
@artemry-nv I think the problem is [this line](https://github.com/mellanox-hpc/jenkins_scripts/blob/1d66874bec127c22a2248b8d0da4cb5c7bdf5158/jenkins/ompi/ompi_test.sh#L228) that needs to be uncommented so that the `XXX_C` (and other) variable is defined for the next test.

**Review by bwbarrett — APPROVED** [review]

**Pull request merged**